src/backend/distributed/planner/local_distributed_join_planner.c

```diff
@@ -0,0 +1,529 @@
+ */
+static bool
+FirstIntListContainsSecondIntList(List *firstIntList, List *secondIntList)
+{
+	int curInt = 0;
+	foreach_int(curInt, secondIntList)
+	{
+		if (!list_member_int(firstIntList, curInt))
+		{
+			return false;
+		}
+	}
+	return true;
+}
+
+
+/*
+ * GetAllUniqueIndexes adds the given index's column numbers if it is a
+ * unique index.
+ * TODO:: if there is a unique index on a multiple column, then we should
+ * probably return true only if all the columns in the index exist in the filter.
+ */
+static void
+GetAllUniqueIndexes(Form_pg_index indexForm, List **uniqueIndexGroups)
+{
+	if (indexForm->indisunique || indexForm->indisprimary)
+	{
+		IndexColumns *indexColumns = palloc0(sizeof(IndexColumns));
+		List *uniqueIndexes = NIL;
+		for (int i = 0; i < indexForm->indkey.dim1; i++)
+		{
+			uniqueIndexes = list_append_unique_int(uniqueIndexes,
+												   indexForm->indkey.values[i]);
+		}
+		if (list_length(uniqueIndexes) == 0)
+		{
+			return;
+		}
+		indexColumns->indexColumnNos = uniqueIndexes;
+		*uniqueIndexGroups = lappend(*uniqueIndexGroups, indexColumns);
+	}
+}
+
+
+/*
+ * RequiredAttrNumbersForRelation returns the required attribute numbers for
+ * the input RTE relation in order for the planning to succeed.
+ *
+ * The function could be optimized by not adding the columns that only appear
+ * WHERE clause as a filter (e.g., not a join clause).
+ */
+static List *
+RequiredAttrNumbersForRelation(RangeTblEntry *rangeTableEntry,
+							   PlannerRestrictionContext *plannerRestrictionContext)
+{
+	RelationRestriction *relationRestriction =
+		RelationRestrictionForRelation(rangeTableEntry, plannerRestrictionContext);
+
+	if (relationRestriction == NULL)
+	{
+		return NIL;
+	}
+
+	PlannerInfo *plannerInfo = relationRestriction->plannerInfo;
+	Query *queryToProcess = plannerInfo->parse;
+	int rteIndex = relationRestriction->index;
+
+	List *allVarsInQuery = pull_vars_of_level((Node *) queryToProcess, 0);
+
+	List *requiredAttrNumbers = NIL;
+
+	Var *var = NULL;
+	foreach_ptr(var, allVarsInQuery)
+	{
+		if (var->varno == rteIndex)
+		{
+			requiredAttrNumbers = list_append_unique_int(requiredAttrNumbers,
+														 var->varattno);
+		}
+	}
+
+	return requiredAttrNumbers;
+}
+
+
+/*
+ * CreateConversionCandidates creates the conversion candidates that might
+ * be converted to a subquery so that citus planners can work.
+ */
+static ConversionCandidates *
+CreateConversionCandidates(PlannerRestrictionContext *plannerRestrictionContext,
+						   List *rangeTableList, Oid resultRelationId)
+{
+	ConversionCandidates *conversionCandidates =
+		palloc0(sizeof(ConversionCandidates));
+
+	RangeTblEntry *rangeTableEntry = NULL;
+	foreach_ptr(rangeTableEntry, rangeTableList)
+	{
+		/* we're only interested in tables */
+		if (!IsRecursivelyPlannableRelation(rangeTableEntry))
+		{
+			continue;
+		}
+
+		/* result relation cannot converted to a subquery */
+		if (resultRelationId == rangeTableEntry->relid)
```

Comment:
I guess here we should check this with an rteIdentity.

Because now 
```sql
UPDATE local l1 SET a = 5 FROM local l2 JOIN dist USING(a);
```
Doesn't consider the local table(l2) as a candidate, because it has the same relid as the result relation but it is not the result relation.

Reply:
This would also work with the current version because it would become router plannable after `dist` is recursivelly planned